src/fsharp/TypeChecker.fs

```diff
@@ -8297,30 +8297,70 @@ and delayRest rest mPrior delayed =
 //------------------------------------------------------------------------- 
 
 and TcFunctionApplicationThen cenv overallTy env tpenv mExprAndArg expr exprty (synArg: SynExpr) atomicFlag delayed = 
-    
     let denv = env.DisplayEnv
     let mArg = synArg.Range
     let mFunExpr = expr.Range
+
+    /// Finds last ident of LongIdent in SynExpr.Ident, LongIdent or TypeApp.
+    /// Type checkes the whole thing as it goes in order to:
+    ///
+    /// * ensure we pass well typed things to `nameof`
+    ///
+    /// * not to loose `FSharpSymbolUse` for `nameof` argument, because we erase it with `Expr.Const(Const.String ...)` further in this function.
+    let (|LastPartOfLongIdentStripParens|_|) expr =
+        let rec findIdents expr (isTypeApp : bool) =
+            match expr with
+            | SynExpr.Ident ident ->
+                if not isTypeApp then
+                    TcExprOfUnknownType cenv env tpenv expr |> ignore
+                Some ident
+            | SynExpr.TypeApp (expr, _, typeArgs, _, _, mTypeArgs, m) -> 
+                TcExprOfUnknownTypeThen cenv env tpenv expr [ DelayedTypeApp (typeArgs, mTypeArgs, m) ] |> ignore
```

Comment:
@dsyme It does not work as expected. What I'm trying to do here is to type check the argument part of expression line this:
```fsharp
nameof System.Collections.Generic.Dictionary<int, System.Collections.Generic.List<string>>
```
It refuses to be checked as expression because it's not a valid expression, it guesses it's a constructor call and wants arguments.